crates/oxc_ast/src/trivia.rs

```diff
@@ -8,19 +8,39 @@ use std::{
 
 use oxc_span::Span;
 
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum CommentPosition {
+    Leading,
+    Trailing,
+    Dangling,
+}
+
 /// Single or multiline comment
 #[derive(Debug, Clone, Copy)]
 pub struct Comment {
     pub kind: CommentKind,
     /// The span of the comment text (without leading/trailing delimiters).
     pub span: Span,
+
+    pub position: CommentPosition,
+    pub preceded_by_newline: Option<bool>,
+    pub followed_by_newline: Option<bool>,
+
+    pub attached_to: u32,
 }
```

Comment:
Need to study what information is required by prettier, cc @leaysgur 

I don't remember where this came from https://github.com/oxc-project/oxc/blob/91bb06e6ff519115f86fab1695a738ab58b875d0/crates/oxc_prettier/src/comments/mod.rs#L54-L65

Reply:
I'm still on the way to read through these area, but as far as I can tell are:

- Prettier performs 3 phases to collect comment information
  - 1) Find `preceding`, `enclosing` and|or `following` node for each comment
    - Relatively intuitive, but there is exception
  - 2) Classify comments into `ownLine`, `endOfLine` and `remaining` group
  - 3) Attach comments to each node with 1 + 2
    - This one using so many utils...
- Final AST nodes contain `comments` field which are `ESTree.Comment[]` and `Comment` is extended with `leading` and `trailing` flag.
  - I found a few more extra properties but not understand yet

😅

Reply:
> I don't remember where this came from 

Maybe this?

https://github.com/prettier/prettier/blob/aa3853b7765645b3f3d8a76e41cf6d70b93c01fd/src/language-js/utils/index.js#L1005